### ISSUE TITLE
error when opening the built app on MacOSX: LSOpenURLsWithRole() failed with error -10810 for the file

### ISSUE THREAD

**Issue by rcholic** [asticode/go-astilectron#103]
I used the go-astilectron demo project (file system exploration) as a template to make some changes, which include a go HTTP webserver. My unit tests are working fine, and the build works okay as well, but when I tried to open the built app from terminal, I got the following errors:

```
LSOpenURLsWithRole() failed with error -10810 for the file /Users/xxxDesktop/Code/Go/Workspace/src/instant_share_app/output/darwin-amd64/Astilectron demo.app.
```
I wonder how to debug this error?

**Comment by asticode** [comment]
- I gather you're working on OSX. Which version?
- is this working when you double click on the app icon?
- I've found a fix for a somewhat similar error: `I ran into this, and found that the error went away if I dragged the old version of the app to the trash (via the Finder) before installing the new version. (I don't know if this works for everybody.)`. Could you try it and tell me whether it's working?

It seems to be a MacOSX bug unfortunately :(

**Comment by rcholic** [comment]
@asticode 
thanks for your help, but my issue was different than yours, because I was able to do some analysis by logging the errors (from my go code) to a text file. The log is very helpful for me to debug the issue that has something to do with file paths in the built app. Now everything is fine. Thanks again!

**Comment by rcholic** [comment]
@asticode 
The error as I reported in my first post is only the consequence of errors that happen in my Go code, and I had no clue as to what caused the errors. So I saved the logging messages to a file on my local disk, e.g.

```go
file, _ := os.Create("/Users/rcholic/Desktop/error_logging.txt")
defer file.Close()
if err := db.Connect("./data/my.db); err != nil {
   file.WriteString(err.Error())
   file.Sync()
}
```
When the app crashed, I was able to see the error in my logging file, which helped me to locate the source of the errors in my code.

The root cause for the errors are file path issues with template html files and my database file. For example:
```go
db, err := Bolt.open("./data/my_local.bolt.db") // the path to my bolt database is relative here
```

When the mac app is built, the app crashed due to the relative path problem, because the app does not know where the heck it is really located. To solve this problem, **I use the following method to locate where the app is being executed:**
```go
        ex, err := os.Executable() // ex is the path to where the app is being executed
	if err == nil {
		dataPath := filepath.Join(filepath.Dir(ex), "my_db_name.bolt.db")
		db, err := services.NewDB(dataPath, true, true)
		defer db.Close()
		if err != nil {
			logrus.Errorf("err in NewDB: %v", err)
		}
	}
```
Hopefully this answer will be helpful to other people using this cool framework **go-astilectron** 😄

**Comment by asticode** [comment]
thanks a lot for sharing this :)

Cheers

**Comment by tpoxa** [comment]
I have same error running Astilectron Demo App. Weird thing is that when I build on my mac directly- its working
When I build using gitlab shell runner, resulted .app throws me this error.

```
build:darwin:amd64:
  stage: build
  tags:
    - macbook
  artifacts:
    expire_in: 1 week
    paths:
      - output
  script:
    - go get -u github.com/asticode/go-astilectron-bundler/...
    - go install github.com/asticode/go-astilectron-bundler/astilectron-bundler  
    - astilectron-bundler
  before_script:
    - export GOROOT="/usr/local/go"
    - export GOPATH="$HOME/go"
    - export PATH="$GOPATH/bin:$GOROOT/bin:$PATH"

```

BTW Build passes with no errors. Only problem that app did not start. Running it with cli gives me an error 
```LSOpenURLsWithRole() failed with error -10810 for the file /Users/username/Downloads/output 4/darwin-amd64/Astilectron demo.app.```

**Comment by tpoxa** [comment]
When I do run binary directly using CLI app is starting and running good.

```
 cd darwin-amd64/Astilectron\ demo.app
 cd Contents/cd MacOS
./Astilectron\ demo
2020/03/10 13:07:33 Running app built at 2020-03-10 12:45:52.322617 +0200 EET m=+4.056817674
2020/03/10 13:07:33 astikit: starting worker...
```
After first run double click is working now.

Mac OS Catalina... weird things.

**Comment by tpoxa** [comment]
Well, for sure I can say it doesn't connected with actually build. As soon as archive was downloaded from internet, app inside it doesn't start.

**Comment by tomachinz** [comment]
I also experience this painful error, hoping to see my home folder appear in the finder:

~ open ../
LSOpenURLsWithRole() failed with error -54 for the file /Users.

This all started when I used RCDefaults to set BBEdit as the default app to open folders, believe it or not! This was instantly regretted, but I found no way to fix except to throw BBEdit in the trash! Then once again Finder would open it. Until I trashed RCDefaults first! Then when trashing BBEdit, Finder would not open folders anymore (using open at CLI). Instead I get that error. 

Only fix for me was to bin BBEdit. 
Nothing to do with Go Astilectron mind. Could be related to default apps, file extensions, MIME types etc.
